File: Assets/Scripts/UI/UIBase.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;
using System.Reflection;
using System;

namespace Kasug
{
    [System.Serializable]
    /// <summary>
    /// Bounds用于点击检测
    /// </summary>
    public struct UIBaseBound
    {
        public Image image;
        public List<Vector2> points;
    }
    public class UIBase : MonoBehaviour
    {
        public int Depth;
        public UIType uiType;

        [SerializeField]
        protected List<UIBaseBound> allBounds;
        public List<UIBaseBound> AllBounds
        {
            get
            {
                if (allBounds == null)
                    allBounds = new List<UIBaseBound>();

                return allBounds;
            }
        }

        public List<UIBaseBound> firstQuadrantUI  = new List<UIBaseBound>();
        public List<UIBaseBound> secondQuadrantUI = new List<UIBaseBound>();
        public List<UIBaseBound> thirdQuadrantUI  = new List<UIBaseBound>();
        public List<UIBaseBound> fourthQuadrantUI = new List<UIBaseBound>();

        public Dictionary<string, TouchClickHandler> UIClick    { private set; get; } = new Dictionary<string, TouchClickHandler>();
        public Dictionary<string, TouchPressHandler> UIPress    { private set; get; } = new Dictionary<string, TouchPressHandler>();
        public Dictionary<string, TouchDragHandler>  UIDrag     { private set; get; } = new Dictionary<string, TouchDragHandler>();

        public Dictionary<string, float> UIPressTime { private set; get; } = new Dictionary<string, float>();

        protected virtual void Awake()
        {
            Image[] childrenImages = GetComponentsInChildren<Image>(true);
            if (childrenImages[0].name.Equals(transform.name))
            {
                Debug.LogErrorFormat("{0}根节点请不要添加Image组件!", transform.name);
                return;
            }

            float middleX = Screen.width * 0.5f;
            float middleY = Screen.height * 0.5f;

            foreach (var img in childrenImages)
            {
                RectTransform rectTransform = img.rectTransform;
                Vector2 pos = UIManager.UICamera.WorldToScreenPoint(rectTransform.position);
                Rect rect = rectTransform.rect;
                Vector4 size = new Vector4(rect.xMin, rect.xMax, rect.yMin, rect.yMax);

                UIBaseBound bound = new UIBaseBound()
                {
                    image = img,
                    points = new List<Vector2>()
                };

                float left = pos.x + size.x;
                float right = pos.x + size.y;
                float bottom = pos.y + size.z;
                float up = pos.y + size.w;
                Vector2 bottomLeft  = new Vector2(left, bottom);
                Vector2 upperLeft = new Vector2(left, up);
                Vector2 upperRight = new Vector2(right, up);
                Vector2 bottomRight = new Vector2(right, bottom);

                bound.points.Add(bottomLeft);
                bound.points.Add(upperLeft);
                bound.points.Add(upperRight);
                bound.points.Add(bottomRight);
                //Transform canvas = GameObject.Find("Canvas").transform;
                //for (int i = 0; i < bound.points.Count; i++)
                //{
                //    GameObject g = new GameObject();
                //    g.name = (i + 1).ToString();
                //    Image iii = g.AddComponent<Image>();
                //    iii.color = Color.red;
                //    g.transform.parent = canvas;
                //    g.transform.localScale = Vector3.one * 0.1f;
                //    g.transform.position = bound.points[i];
                //}
                allBounds.Add(bound);

                JudgeQuadrant(bound, middleX, middleY);
            }
        }

        protected virtual void Start()
        {
            string className = uiType.ToString() + "_V";
            Type type = Type.GetType(className);

            if(type == null)
            {
                Debug.LogError("没有找到" + className + "对应类");
                return;
            }

            gameObject.AddComponent(type);
        }

#if UNITY_EDITOR
        public bool showBound = false;
        [Range(1f,50f)]
        public float boundsScale = 20f;
        private void OnDrawGizmos()
        {
            Gizmos.color = Color.red;
            if (showBound)
            {
                foreach (var p in AllBounds)
                {
                    foreach (var point in p.points)
                    {
                        Gizmos.DrawWireSphere(point, boundsScale);
                    }
                }
            }
        }
#endif

        

        protected void AddUiButton(string buttonName,TouchClickHandler clickEvent)
        {
            UIClick.Add(buttonName, clickEvent);
        }

        protected void AddUiDrag(string buttonName, TouchDragHandler dragEvent)
        {
            UIDrag.Add(buttonName, dragEvent);
        }

        protected void AddUiPress(string buttonName, TouchPressHandler pressEvent,float pressTime = 1f)
        {
            UIPressTime.Add(buttonName, pressTime);
            UIPress.Add(buttonName, pressEvent);

        }


        #region Utils

        private void JudgeQuadrant(UIBaseBound bound,float middleX,float middleY)
        {
            foreach (var p in bound.points)
            {
                if (p.x >= middleX && p.y >= middleY)
                {
                    Add(firstQuadrantUI, bound);
                }
                else if (p.x >= middleX && p.y < middleY)
                {
                    Add(fourthQuadrantUI, bound);
                }
                else if (p.x < middleX && p.y >= middleY)
                {
                    Add(secondQuadrantUI, bound);
                }
                else if (p.x < middleX && p.y < middleY)
                {
                    Add(thirdQuadrantUI, bound);
                }
            }

        }

        private void Add(List<UIBaseBound> list, UIBaseBound bound)
        {
            if (list.Contains(bound)) return;
            list.Add(bound);
        }

        #endregion
    }
}

File: Assets/Scripts/ui/UIBase.lua
local UIBase = class("UIBase")
local UIButton = require("ui.UIButton")

local UIClick       = {}
local UIDrag        = {}
local UIPress       = {}
local UIPressTime   = {}

--四个象限里存 UIBaseBound
local firstQuadrantUI = {}
local secondQuadrantUI = {}
local thirdQuadrantUI = {}
local fourthQuadrantUI = {}

local Camera = require("world.Camera")

function UIBase:OnShow(data)    

    self.rectTransform.anchorMin = Vector2.zero
    self.rectTransform.anchorMax = Vector2.one
    self.rectTransform.anchoredPosition3D = Vector3(0,0,-self.rectTransform.parent.anchoredPosition3D.z)
    self.rectTransform.offsetMin = Vector2.zero
    self.rectTransform.offsetMax = Vector2.zero
end

function UIBase.RemoveAllButtonListeners(self)
    dump(self)
end

-- function UIBase:OnDestroy()
--     dump(self, "=====")
--      self:RemoveAllButtonListeners()
-- end

     --[[
    --派发事件必须
    Component(self)
        :addComponent("components.behavior.EventProtocol")
        :exportMethods()

    local transform             = self.gameObject:GetComponent(typeof(CS.UnityEngine.RectTransform));
    transform.anchorMin         = Vector2.zero;
    transform.anchorMax         = Vector2.one;
    transform.anchoredPosition  = Vector2.zero;
    transform.sizeDelta         = Vector2.zero;

    local uiButtons             = self.gameObject:GetComponentsInChildren(typeof(UIButton),true)
    local middleX               = CS.UnityEngine.Screen.width * 0.5
    local middleY               = CS.UnityEngine.Screen.height * 0.5

    local uiCamera              = Camera:GetInstance().UICamera;

    --计算四个角点
    for _,v in ipairs(uiButtons) do
        local rectTransform = v.rectTransform
        local pos               = uiCamera:WorldToScreenPoint(rectTransform.position)
        local rect              = rectTransform.rect
        local size              = {rect.xMin,rect.xMax,rect.yMin,rect.yMax}
        local bound             = {}
        local image             = v.sourceImage

        local left              = pos.x + size[1]
        local right             = pos.x + size[2]
        local bottom            = pos.y + size[3]
        local up                = pos.y + size[4]

        local bottomLeft        = Vector2(left, bottom)
        local upperLeft         = Vector2(left, up)
        local upperRight        = Vector2(right, up)
        local bottomRight       = Vector2(right, bottom)

        bound[1][image]         = image

        bound[2][1]             = bottomLeft
        bound[2][2]             = upperLeft
        bound[2][3]             = upperRight
        bound[2][4]             = bottomRight

        JudgeQuandrant(bound)
    end
end

--判断象限
function JudgeQuandrant(bound,middleX,middleY)
    for k,p in ipairs(bound) do 
        if p.x >= middleX & p.y >= middleY then
            table.insert(firstQuadrantUI,bound)
        elseif p.x >= middleX & p.y < middleY then
            table.insert(fourthQuadrantUI,bound)            
        elseif p.x < middleX & p.y >= middleY then            
            table.insert(secondQuadrantUI,bound)
        elseif p.x < middleX & p.y < middleY then            
            table.insert(thirdQuadrantUI,bound)
        end
    end
end--]]
function SendProto(protoName,args,successEvent)
    utils:StartCoroutine(function() 
        local future, data = network:SendAndWait("ETModel."..protoName, args)
        if future:isRejected() then         
            Debug.LogError("ETModel."..protoName.." failed");
            return;
        end
        successEvent()
    end)
end

return UIBase

